src/main/java/com/aws/greengrass/mqttbridge/clients/IoTCoreClient.java

```diff
@@ -109,7 +146,11 @@ private void publishToIotCore(String topic, byte[] payload) {
 
     @Override
     public synchronized void updateSubscriptions(Set<String> topics, @NonNull Consumer<Message> messageHandler) {
+        if (subscribeFuture != null) {
```

Comment:
Again, you have to synchronize this _on the same lock_.

Use the `subscribeLock` to synchronize here.

Reply:
Synchronization must include the `subscribeFuture = ` at the end

Reply:
that method is already synchronized, right?

Reply:
What does the `synchronized` keyword do in a function signature? It synchronizes on `this`. 

`this` is not the same object as `subscribeFuture`, right? Therefore, you are not synchronizing on the same object. Therefore, you have a race and a bug.